docs/design/diagnostics.md

```diff
@@ -0,0 +1,293 @@
+- The ability to deduplicate or manipulate diagnostics.
+
+## Proposals
+
+### Thin Abstraction
+
+The most trivial implementation would be creating new Go types in this framework that provide a thin abstraction layer, e.g.
+
+```go
+package diag
+
+type Diagnostic struct {
+    Severity  DiagnosticSeverity
+    Summary   string
+    Detail    string
+    Attribute *tftypes.AttributePath // See also: https://github.com/hashicorp/terraform-plugin-framework/issues/81
+}
+
+type Diagnostics []Diagnostic
+
+func (d Diagnostics) Append(in ...Diagnostic)
+func (d Diagnostics) HasError() bool
+func (d Diagnostics) toTfprotov6Diagnostics() []*tfprotov6.Diagnostic
+```
+
+This solves many requirements, allowing the framework to own the abstraction and provide ergonomic methods. However, this implementation is tightly coupled to the underlying implementation and does not introduce protections against backwards compatibility issues. Since the data fields are strongly associated with the type, implementors inside and outside the framework are bound to the initial implementation details. If those details are not exported and instead hidden behind a constructor function, e.g.
+
+```go
+type Diagnostic struct {
+    severity  DiagnosticSeverity
+    summary   string
+    detail    string
+    attribute *tftypes.AttributePath // See also: https://github.com/hashicorp/terraform-plugin-framework/issues/81
+}
+
+func NewDiagnostic(severity DiagnosticSeverity, summary string, detail string, attribute *tftypes.AttributePath) Diagnostic
+func NewErrorDiagnostic(summary string, detail string, attribute *tftypes.AttributePath) Diagnostic
+func NewWarningDiagnostic(summary string, detail string, attribute *tftypes.AttributePath) Diagnostic
+```
+
+This type of implementation will still be difficult to extend or change.
+
+As another extensibility example, as `Diagnostics` pass through various functions across subsystems, determining the "source" or "cause" of a diagnostic must be done by `Summary` and/or `Detail` value similar to checking Go `error` types by value using `(error).Is()`. Extending the `Diagnostic` type is possible to add other identifying fields or creating new types, e.g.
+
+```go
+type ValidationDiagnostic struct {
+  Diagnostic
+}
+```
+
+However, the implementation for diagnostics then differs from other design choices in the framework where interfaces are preferred. For example, it becomes difficult to implement equality and future enhancements for diagnostics since the interface is not strongly defined for extensions.
+
+### Basic Interface
+
+Diagnostics could be modeled around interfaces:
+
+```go
+package diag
+
+type Diagnostic interface {
+  Severity() Severity
+  Summary()  string
+  Detail()   string
+
+  Equal(other Diagnostic) bool
+}
+
+type DiagnosticWithLogger interface {
+  Diagnostic
+
+  Log(ctx context.Context)
+}
```

Comment:
Unclear on what this interface is or what it would do.

Reply:
It's meant as an example extension, where the framework could support writing a log entry matching a diagnostic. Will add code comments, or happy to show a different example.